### DISCUSSION TITLE
Let's discuss about `asciinema edit`

### DISCUSSION THREAD

**Discussion by s-leroux** [asciinema/discussions#66] [Ideas]
Hi everyone,

We regularly feel the need for editing asciicast. There are quite a few issues raised here, and even a few pull requests:

* https://github.com/asciinema/discussions/issues/54
* https://github.com/asciinema/asciinema/pull/339
* https://github.com/asciinema/discussions/issues/53
* https://github.com/asciinema/discussions/issues/38
* https://github.com/asciinema/asciinema/issues/138
* https://github.com/asciinema/asciinema/issues/19

For my own needs, a few years ago, I added ad-hoc editing capabilities to the version of asciinema available on my system, partly by hacking the source, partly by writing an external tool. And I am probably not the only one in that case. But this is now a very old version, and since I will have a bit of spare time in the next few weeks, maybe it's a good time to contribute to an _official_ solution. However, before jumping into the code, I think it would be great to discuss more in-depth about the editing features we want/need. That's the purpose of this "thread".

## Features
At the very least, I think we should provide the following editing capabilities:

1. **Delete a part of an asciicast.** At first sight, that shouldn't be hard, since we just have to discard the corresponding events
2. **Extract a part of an asciicast.** Once again, it shouldn't be too hard.
3. **Join _2_ asciicasts.** 

Potentially more complex, but required I think:

4. **Squash a range of events**. Either by merging them or by making them happen simultaneously (or the closest possible to that). See comment by Marcin here: https://github.com/asciinema/asciinema/pull/339#issuecomment-473699217
5. **Limit delays between events** (failing to find a better wording here) The `rec` command has the `--idle-time-limit` option. But it would be nice if we can reduce idle time in post-processing (no one wants to see you typing so sloooowly). I also read somewhere the need for a fast-forward playback. Maybe we could add a _speed up/slow down_ option as part of this feature.
6. **insert delay**. This is somewhat the opposite of the previous feature. I sometimes need to sync asciicast with voiceover. Maybe this is a very specific need, but that would be of great help if I could insert a pause or delay in an asciicast.
7. **Hide sensitive content**. I think we need a way to replace the characters in a range of input/output events by stars or bullets or something else to hide sensitive data.

## Indexing events
Most, if not all, the features above need a way to specify the part of the asciicast where they apply. It is not hard to imagine `--from` and `--to` options for that.  The first idea would be to specify the start and end positions using a timecode-like format `mm:ss.___`. But for very precise work, I think we should also be able to specify a position by an event number.

## UI
From the end-user perspective, I see two ways the `edit` command could be implemented:

* Either we only allow one basic editing sub-command for each invocation of  `asciinema edit`.
* or (somewhat like ImageMagick `convert`) we can chain multiple sub-commands in the same invocation of `asciinema edit`.

The first solution seems simpler to implement. On the other hand, a basic task like joining three asciicasts and inserting a pause between each segment would require running several commands sequentially, eventually requiring the user to keep track of the different timecodes or segment duration between the commands.

At this time, I really don't know where to put the cursor between code complexity and ease of use. 

## Your turn now!
Those are only raw ideas I had while thinking about an editing command for asciinema. But I don't have a deep knowledge of the project. And I can't tell if some of these ideas would be overly complex to implement. Finally, _you_ may have your own ideas about this topic or you may have identified needs I failed to address. So don't hesitate to comment below!

**Comment by mplattner** [comment]
FTR: With https://github.com/cirocosta/asciinema-edit you can cut existing casts or speed them up/down.

**Comment by johnandersen777** [comment]
It seems like editing would be easier if each record in the NDJSON recorded the delta since the last record rather than an absolute timestamp. Thoughts on if this a good or flawed approach?

**Reply by cprima** [reply]
Indeed, when I just set out to write me a quick Python script to truncate a .cast file that was indeed my approach:
- 1st pass: for each line calculate the delta from the timestamp of the previous line
- 2nd pass: truncate by a list of start/end timestamps and update the timestamps after a truncation by adding the previous line's timestamp + the current line's delta.

Here is a gist: https://gist.github.com/cprima/d0973019db4d0654a364a1498f45ff80

I had a asciinema recording of an installation process which showed for 10 minutes nothing but a progress-bar-like screen.
Trauncated that easily by calling` truncate_asciinema("long.cast", "truncated.cast", [(30, 630), (715, 725)])`

**Reply by ku1ik** [reply]
The new [asciicast v3 format](https://docs.asciinema.org/manual/asciicast/v3/) uses intervals (delta). This should make editing easier (for both humans and tools).
